### PR TITLE
docs: Add a first draft of the Go SDK tech preview docs #3360

### LINKED ISSUE #3177: Create FAQ page
Add common questions

### LINKED ISSUE #3360: [EPIC] Go SDK Doc Launch Plan
Docs Title:  Go SDK for Dagger  
Note - it needs to have a tech preview banner

**Sidebar plan:** 
- Introduction 
- Install page 
- Go Getting Started
- API Reference
- FAQ

**URL plan and where it lives:** 
Build in the current site - @vikram-dagger I'll let you create an issue for this since I am not sure if we just need a PR instead of an issue. 

**What pages do we need for launch:** 
- [ ] Introduction #3361  - use this content and tweak to focus on SDK theme - https://github.com/dagger/dagger/tree/cloak
@shykes  will provide guidance and reviews here when needed 
- [ ] Installation page #3362 - Make sure it has min. version requirements 
- [ ] Go getting started guide #3363 - Make sure to include install instructions for Go SDK (point to install page above) 
@kpenfound  help support Vikram here 
- [ ] FAQ #3177  (Vikram’s questions and we will expand later) 
Must have a “what’s up with CUE?” question - asked @shykes  to add the content for that one. 
- [ ] Go SKD reference page #3364 - Just link to Go generated docs 
@vito  will lead proposal, but @vikram-dagger will implement. We will look into hosting ourselves later (#3365).

**User experience summary:** 
This is the experience that we are targeting for launch. 

**Scenario 1-** 

- Sees banner about blog post on website or social 
- Clicks docs link in blog post 
- Goes to Go SDK docs 

**Scenario 2 -** 

- Clicks docs tab on homepage 
- Taken to old version docs, but sees a CTA to check out the new Go SDK (Miranda create a proposal issue) 
- Clicks SDK CTA and goes to Go docs with no backtracking option to see old docs 

**Additional issues to do to achieve user experience:**  
- [ ] Need pointer to go sdk docs from main docs since it won’t automatically show the new docs when they click docs #3366

**Timeline:** 
- Monday - review PR 
- Tuesday - go/no go decision made 
- Wednesday - merge PR 
- Thursday - launch if PR merges

### LINKED ISSUE #3361: Introduction for Go SDK docs
We need an intro page for the Go SDK docs #3360 

We should use the content from the intro readme here, and alter it to focus on the Go SDK instead of extensions: https://github.com/dagger/dagger/tree/cloak

### LINKED ISSUE #3362: Installation page for Go SDK
related to #3360 

need a specific page for installation instructions

### LINKED ISSUE #3363: Go Getting Started Guide
related to #3360 

We need a getting started guide that focuses on the Go SDK. @kpenfound can help support with any Go questions 

It should have a section for installation, but should point back to the install page created in #3362

### LINKED ISSUE #3364: Go SDK Reference page for Go Launch
Referred to #3360 

For now, we will just link to the Go generated docs and think about hosting it ourselves later (#3365

### LINKED ISSUE #3366: CTA for the Go SDK docs from the main docs page
As mentioned in #3360, we aren't going to push the users to the new Go SDK docs automatically. Therefore, we just need some type of CTA within the old version docs that mentioned where they can find the Go SDK docs when needed. 

This can be in the form of a banner or just a basic CTA on the main page that they see when they click on docs. I will let @slumbering and @vikram-dagger make the choice here.

### PR DESCRIPTION
**Pull request by vikram-dagger**
These docs are added in the v0.3 directory.

After this change, the docs will contain:
- Introduction for Go SDK docs #3361
- Installation instructions #3362
- Get started guide #3363
- FAQ #3177
- Link to auto-generated SDK reference docs #3364

Closes #3360
Closes #3361
Closes #3362
Closes #3363
Closes #3177
Closes #3364

Signed-off-by: Vikram Vaswani <vikram@dagger.io>

### CODE REVIEW

File: docs/v0.3/partials/_install-sdk-go.md
Comment (vikram-dagger) [inline docs/v0.3/partials/_install-sdk-go.md:8]: This is implemented as a partial so it can be reused. Maybe we can add it to our release process so this always reflects the latest tag.

File: docs/v0.3/partials/_install-sdk-go.md
Comment (vikram-dagger) [inline docs/v0.3/partials/_install-sdk-go.md:10]: I'm not sure if this is part of the installation steps or not. I had difficulty running the tutorial steps without it. TBD

File: docs/v0.3/sdk/go/fssrz-index.md
Comment (vikram-dagger) [inline docs/v0.3/sdk/go/fssrz-index.md:2]: This is the URL for the landing page of the Go SDK tech preview docs. Please lmk if it needs to be changed.

File: docs/v0.3/sdk/go/fssrz-index.md
Comment (vikram-dagger) [inline docs/v0.3/sdk/go/fssrz-index.md:24]: I have modified Solomon's original diagram slightly here. Please let me know if further changes are needed.

File: docs/v0.3/sdk/go/xfgss-faq.md
Comment (vikram-dagger) [inline docs/v0.3/sdk/go/xfgss-faq.md:5]: Please review these questions and answers and let me know if we need to add or update them. 

Note that this FAQ is presented at the root level in the navigation tree and not under the "Dagger SDK for Go" node because the questions we have to date seem to be more about Dagger 0.3 than specifically about the Go SDK.

File: docs/v0.3/sdk/go/fssrz-index.md
Comment (vikram-dagger) [inline docs/v0.3/sdk/go/fssrz-index.md:5]: We have previously agreed that this would be called "Go SDK for Dagger". However the convention seems to be "PRODUCT SDK for Go" rather than the other way around. Therefore I have changed it accordingly. Please let me know if you prefer the original version.

File: docs/v0.3/sdk/go/xfgss-faq.md
Comment (vikram-dagger) [inline docs/v0.3/sdk/go/xfgss-faq.md:15]: I've kept mention of GraphQL to a minimum, and only in this FAQ. Please let me know if we need to make this more visible.

File: website/docusaurus.config.js
Comment (vikram-dagger) [inline website/docusaurus.config.js:127]: This plugin is required when embedding partials into markdown.

File: docs/v0.3/partials/_install-sdk-go.md
Comment (kpenfound) [reply docs/v0.3/partials/_install-sdk-go.md:8]: I think once we get to GA (or whenever the Go SDK gets a stable release), we can remove the version specifier since the latest tag will be automatically pulled
https://go.dev/doc/modules/release-workflow#pre-release

**Comment by vikram-dagger** [comment]
It's important to note that there are two places which link specifically to the version tag of the SDK:
- the installation instructions
- the sidebar link to the auto-generated package documentation

We may want to consider adjusting these to be more generic (in the case of the link at least) or including a step in our SDK release process to update these tags

**Comment by mircubed** [comment]
Note to self - we should add a Q&A about where they should go to ask for help or support. I'll add it once this PR gets merged

File: docs/v0.3/sdk/go/fssrz-index.md
Comment (gerhard) [reply docs/v0.3/sdk/go/fssrz-index.md:2]: Looks good to me. WDYT @shykes?

File: docs/v0.3/sdk/go/fssrz-index.md
Comment (gerhard) [reply docs/v0.3/sdk/go/fssrz-index.md:5]: Definitely one for @shykes.

**Comment by vikram-dagger** [comment]
> Note to self - we should add a Q&A about where they should go to ask for help or support. I'll add it once this PR gets merged

@mircubed if you have this information already, I can add it in my next revision.

**Comment by vikram-dagger** [comment]
I believe the linter failure is not justified in this case. There is no code block on line 183. The linter appears to be incorrectly identifying the second level of list nesting (4 spaces) as a code block (also usually 4 spaces). This error has occurred in other PRs as well. If others agree, we should consider disabling linter rule MD046.

**Comment by mircubed** [comment]
> > Note to self - we should add a Q&A about where they should go to ask for help or support. I'll add it once this PR gets merged
> 
> @mircubed if you have this information already, I can add it in my next revision.

Here you go: 

Q - I am stuck. How can I get help? 
A: Join us on [Discord](https://discord.com/invite/dagger-io), and ask your question in our [help forum](https://discord.com/channels/707636530424053791/1030538312508776540). Our team will be happy to help you there!

**Comment by vikram-dagger** [comment]
> Q - I am stuck. How can I get help? A: Join us on [Discord](https://discord.com/invite/dagger-io), and ask your question in our [help forum](https://discord.com/channels/707636530424053791/1030538312508776540). Our team will be happy to help you there!

Added

**Comment by shykes** [comment]
Thanks @vikram-dagger ! What's the best way to review them in rendered form?

File: docs/v0.3/sdk/go/fssrz-index.md
Comment (shykes) [reply docs/v0.3/sdk/go/fssrz-index.md:5]: Good question...

* No matter what, when the context is unambiguously Dagger-specific, we should use the short form "Go SDK". For example in our navbar.

* I'm not sure what the correct long form is. But Stripe has a [Stripe iOS SDK](https://stripe.dev/stripe-ios/docs/) mentions, maybe we could have a "Dagger Go SDK" ?

File: docs/v0.3/partials/_install-sdk-go.md
Comment (shykes) [inline docs/v0.3/partials/_install-sdk-go.md:8]: Since we don't want to version this page, should we consider removing the version-specific part?

In other words:

```suggestion
go get go.dagger.io/dagger
```

File: docs/v0.3/sdk/go/fssrz-index.md
Comment (shykes) [inline docs/v0.3/sdk/go/fssrz-index.md:5]:
```suggestion
# Dagger Go SDK
```

Per discussion: let's give this a try if you agree.

File: website/sidebars.js
Comment (shykes) [inline website/sidebars.js:16]:
```suggestion
      label: "Go SDK",
```

Short form, per discussion

File: docs/v0.3/sdk/go/fssrz-index.md
Comment (shykes) [reply docs/v0.3/sdk/go/fssrz-index.md:24]: A few important details:

1. Let's use "tool" instead of "application", will be less confusing in context (there is probably a Go application nearby, but we don't touch it)
2. Make the diagram specifically about this SDK ("your go tool", "go SDK"...)
3. In this first launch, the engine will not yet be released as a standalone component, therefore we are keeping it out of the spotlight for now. Better to have a simplified diagram that matches the simplified story, and gradually add detail as we launch more pieces.

File: docs/v0.3/sdk/go/8g34z-get-started.md
Comment (shykes) [inline docs/v0.3/sdk/go/8g34z-get-started.md:28]:
```suggestion
## Step 1: Create a simple Go tool
```

Important to talk about "tool" instead of "application". To avoid confusion with the actual application, which is probably being built, tested and deployed by the tool.

File: docs/v0.3/sdk/go/8g34z-get-started.md
Comment (shykes) [inline docs/v0.3/sdk/go/8g34z-get-started.md:31]:
```suggestion
If you already have a Go tool, you can [proceed to Step 2](#step-2-install-the-dagger-sdk-for-go).
```

File: docs/v0.3/sdk/go/8g34z-get-started.md
Comment (shykes) [inline docs/v0.3/sdk/go/8g34z-get-started.md:34]:
```suggestion
Create a simple Go tool as follows:
```

File: docs/v0.3/sdk/go/8g34z-get-started.md
Comment (shykes) [inline docs/v0.3/sdk/go/8g34z-get-started.md:84]:
```suggestion
1. Run the tool by executing the command below.
```

File: docs/v0.3/sdk/go/8g34z-get-started.md
Comment (shykes) [inline docs/v0.3/sdk/go/8g34z-get-started.md:90]:
```suggestion
  The tool displays the string `Hello`.
```

File: docs/v0.3/sdk/go/8g34z-get-started.md
Comment (shykes) [inline docs/v0.3/sdk/go/8g34z-get-started.md:96]:
```suggestion
## Step 3: Run the tool with Dagger
```

Comment (shykes) [inline docs/v0.3/sdk/go/8g34z-get-started.md:96]:
```suggestion
## Step 3: add a pipeline to your tool
```

Not sure about the exact wording.

File: docs/v0.3/sdk/go/8g34z-get-started.md
Comment (shykes) [inline docs/v0.3/sdk/go/8g34z-get-started.md:98]: At this point I am pretty confused. My earlier corrections about "tool" vs "application" assumed this guide is helping write my first tool. With perhaps a sample app involved (something to build, test or deploy).

But it appears that both tool and application are mixed together in this guide?

If that's the case, I think we should avoid this. It's confusing to the end user because it doesn't set expectation for the real world, where tool and application will *definitely* be separate.

Sorry if I missed something here.

**Review by shykes — CHANGES_REQUESTED** [review]

File: docs/v0.3/partials/_install-sdk-go.md
Comment (kpenfound) [reply docs/v0.3/partials/_install-sdk-go.md:8]: This wont work until we have a non-prerelease tag, right?

**Review by shykes — CHANGES_REQUESTED** [review]
Thanks @vikram-dagger !

1. On the getting started: let's leave as-is, @kpenfound is working on updating his code, he has some ideas to improve it. Let's keep your last version for now and wait for his update - then resume review and discussions :)

2. Can you remove the changes affecting the *current* docs from this PR, so we can merge it as-is, and continue iterating on top? For example I will start editing the intro text to match my messaging work, etç.

TLDR: if you can make it such that merging this PR doesn't affect the current docs in any way (plus the CI checks pass) then we should merge ASAP and continue iterating in follow-up PRs.

Thanks!

File: docs/v0.3/partials/_install-sdk-go.md
Comment (shykes) [reply docs/v0.3/partials/_install-sdk-go.md]: cc @aluzzardi

**Review by shykes — APPROVED** [review]
Looks good!

**Comment by gerhard** [comment]
Were you expecting these changes to be published @shykes @vikram-dagger? If not, then this behaves as expected:

image

My understanding was that we auto-publish as we are slowly taking the new 0.3 docs live, but are not making them widely public just yet. cc @mircubed

**Comment by shykes** [comment]
Just curious what the system was. Yes we can stick to the current system, it's fine.

**Pull request merged**
